### PR TITLE
Creating a new FPGA reference design: Merge Sort

### PR DESCRIPTION
**Pull request by tyoungsc**
# Adding a New Sample(s)
## Description
This new FPGA reference design is a merge sort design that utilizes the FPGAs spatial computation using thread- and SIMD-level parallelism.

## Checklist
Administrative
- [X] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): @mdbtucker
- [X] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager: @JoeOster

Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [X] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see @JoeOster for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [X] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+This DPC++ reference design demonstrates a highly paramaterizable merge sort algorithm on an FPGA.
+
+***Documentation***:  
+* [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
+* [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
+* [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming. 
+ 
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
```
Comment (JoeOster) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md:11]: Remove Linux or add a comma

**Review by mdbtucker — DISMISSED** [review]

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+
+    std::cout << "Execution time: " << avg_time_ms << " ms\n";
+    std::cout << "Throughput: " << (input_count_mega / (avg_time_ms * 1e-3))
+              << " Melements/s\n";
+
+    std::cout << "PASSED\n";
+    return 0;
+  } else {
+    std::cout << "FAILED\n";
+    return 0;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/main.cpp:251]: The failure return code should be != 0.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/sorting_networks.hpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
+#ifndef __SORTINGNETWORKS_HPP__
+#define __SORTINGNETWORKS_HPP__
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+#include "impu_math.hpp"
+
+#define SWAP(a, b)  \
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/sorting_networks.hpp:9]: Why use a macro rather than an inline function?

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

**Review by tomlenth — DISMISSED** [review]
no issues.  thanks

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/main.cpp (hunk 2 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/main.cpp]: fixed

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    runs = atoi(argv[2]);
+  }
+
+  // get the random number generator seed as the third command line argument
+  if (argc > 3) {
+    seed = atoi(argv[3]);
+  }
+
+  // enforce at least two runs
+  runs = std::max((int)2, runs);
```
[the remark this answers is not in the document]
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/main.cpp]: Made it an error instead.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md (hunk 1 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md]: Done

**Review by mdbtucker — DISMISSED** [review]

**Review by JoeOster — COMMENTED** [review]
I didn't see any third-party components, did I miss any?

**Comment by mdbtucker** [comment]
> I didn't see any third-party components, did I miss any?

No, you didn't miss any, there are no third party components in this design.

**Comment by tyoungsc** [comment]
@pmpeter1 I addressed your three comments. When you get the chance, can you verify that the changes are sufficient?

**Review by pmpeter1 — APPROVED** [review]

**Comment by tyoungsc** [comment]
Sorry @mdbtucker and @pmpeter1: I found two functions which didn't match the proper style guidelines and changed them.

**Review by mdbtucker — APPROVED** [review]

**Pull request merged**